aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -1659,65 +1672,89 @@ AT_ERROR("cholesky: MAGMA library not found in "
     result = at::empty(input.numel() + 1, input.options());
-    result.resize_as_(input).copy_(input).transpose_(-1, -2);
-  } else {
-    result = cloneBatchedColumnMajor(upper ? self.transpose(-1, -2) : self);
+    result.resize_as_(input).transpose_(-2, -1);
+    TORCH_INTERNAL_ASSERT(result.transpose(-2, -1).is_contiguous());
+    result.copy_(input);
   }
 
   AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(
-      self.scalar_type(), "cholesky_cuda", [&] {
-        apply_cholesky<scalar_t>(result, false, infos);
-      });
+    input.scalar_type(), "cholesky_cuda", [&] {
+      apply_cholesky<scalar_t>(result, upper, info);
+    });
 
-  if (self.dim() > 2) {
-    batchCheckErrors(infos, "cholesky_cuda");
-  } else {
-    singleCheckErrors(infos[0], "cholesky_cuda");
+  if (input.dim() > 2) {
+    // if upper=true we need to tranpose and conjugate the result tensor
+    // because the cholesky decomposition is stored in the lower triangular part
+    if (upper) {
+      input.copy_(result.conj().transpose(-2, -1));
+    } else {
+      input.copy_(result);
+    }
   }
-
-  return upper ? result.transpose_(-1, -2) : result;
 }
 
 // Todo: cusolverDnXpotrfBatched has some numerical issue and is not used
 //     here. Batched cholesky is dispatched to magma.
 //     We will switch to cusolverDnXpotrfBatched after the issue is fixed.
 //     See https://github.com/pytorch/pytorch/issues/53879.
 Tensor _cholesky_helper_cuda(const Tensor& self, bool upper) {
+  auto info_shape = IntArrayRef(
+    self.sizes().cbegin(), self.sizes().cend() - 2); // self.shape[:-2]
+  Tensor self_working_copy = cloneBatchedColumnMajor(self);
+  Tensor info = at::empty({info_shape}, self.options().dtype(kInt));
+
 #ifdef USE_CUSOLVER
   if (batchCount(self) == 1 || !use_magma_) {
-    return _cholesky_helper_cuda_cusolver(self, upper);
-  }
-  else {
-    return _cholesky_helper_cuda_magma(self, upper);
+    cholesky_helper_cusolver(self_working_copy, upper, info);
+  } else {
+    cholesky_helper_magma(self_working_copy, upper, info);
   }
 #else
-  return _cholesky_helper_cuda_magma(self, upper);
+  cholesky_helper_magma(self_working_copy, upper, info);
 #endif
+
+  if (self.dim() > 2) {
+    batchCheckErrors(info, "cholesky_cuda");
+  } else {
+    singleCheckErrors(info.item<int64_t>(), "cholesky_cuda");
+  }
+  return self_working_copy;
+}
+
+// The same MAGMA/cuSOLVER dispatch logic is used here as in _cholesky_helper_cuda
+// This function is for cholesky_stub that is used for linalg_cholesky_ex
+// while _cholesky_helper_cuda is used for legacy torch.cholesky
+void cholesky_kernel(const Tensor& input, const Tensor& info) {
+  bool upper = false;
+#ifdef USE_CUSOLVER
+  if (batchCount(input) == 1 || !use_magma_) {
+    cholesky_helper_cusolver(input, upper, info);
+  } else {
+    cholesky_helper_magma(input, upper, info);
+  }
+#else
+  cholesky_helper_magma(input, upper, info);
+#endif // USE_CUSOLVER
```

Comment:
If `cholesky_kernel` and `_cholesky_helper_cuda` will be using the same heuristic, it would be really nice to put them into once place.

Reply:
Yes, but `cholesky_kernel` in its current form doesn't accept `bool upper`. It's easy to change, but it was done on purpose since `torch.linalg.cholesky` doesn't support it and I wanted to keep `torch.cholesky` and `torch.linalg.cholesky` separate, to introduce minimal changes.

Looking at it now I can probably add `bool upper` to `cholesky_kernel` and implement all the cholesky functions using the same `cholesky_stub`. But I should probably split this PR.